girderformindlogger/models/user.py

```diff
@@ -170,34 +170,48 @@ def authenticate(self, login, password, otpToken=None):
         login = login.lower().strip()
         loginField = 'email' if '@' in login else 'login'
 
-        user = self.findOne({loginField: login})
-        if user is None:
+        if loginField=='login':
+            user = self.findOne({loginField: login})
+        else:
+            users = list(self.find({loginField: login}))
+        if user is None and users is None:
             raise AccessException('Login failed.')
 
-        # Handle users with no password
-        if not self.hasPassword(user):
-            e = events.trigger('no_password_login_attempt', {
-                'user': user,
-                'password': password
-            })
+        if user:
+            # Handle users with no password
+            if not self.hasPassword(user):
+                e = events.trigger('no_password_login_attempt', {
+                    'user': user,
+                    'password': password
+                })
 
-            if len(e.responses):
-                return e.responses[-1]
+                if len(e.responses):
+                    return e.responses[-1]
 
-            raise ValidationException(
-                'This user does not have a password. You must log in with an '
-                'external service, or reset your password.')
+                raise ValidationException(
+                    'This user does not have a password. You must log in with an '
+                    'external service, or reset your password.')
 
-        # Handle OTP token concatenation
-        if otpToken is True and self.hasOtpEnabled(user):
-            # Assume the last (typically 6) characters are the OTP, so split at that point
-            otpTokenLength = self._TotpFactory.digits
-            otpToken = password[-otpTokenLength:]
-            password = password[:-otpTokenLength]
+            # Handle OTP token concatenation
+            if otpToken is True and self.hasOtpEnabled(user):
+                # Assume the last (typically 6) characters are the OTP, so split at that point
+                otpTokenLength = self._TotpFactory.digits
+                otpToken = password[-otpTokenLength:]
+                password = password[:-otpTokenLength]
 
-        # Verify password
-        if not self._cryptContext.verify(password, user['salt']):
-            raise AccessException('Login failed.')
+            self._verify_password(password, user)
+
+        else:
+            user = {}
+            verified_password = False
+            for u in users:
```

Comment:
Is ```users``` defined before this? If it's not defined this will be a ```NameError``` and if its defined as ```None``` it will be a ```TypeError```

Reply:
Yeah, it's not superelegant, but https://github.com/ChildMindInstitute/mindlogger-app-backend/blob/ba7fc7fabfc0da3f10661eedb39fd0b5193bf2b2/girderformindlogger/models/user.py#L173-L178 checks if we're using a username or email address. If neither, it raises an exception. If a username, it executes https://github.com/ChildMindInstitute/mindlogger-app-backend/blob/ba7fc7fabfc0da3f10661eedb39fd0b5193bf2b2/girderformindlogger/models/user.py#L180-L202 and skips this `else`; it should only hit this `else` if it also hit https://github.com/ChildMindInstitute/mindlogger-app-backend/blob/ba7fc7fabfc0da3f10661eedb39fd0b5193bf2b2/girderformindlogger/models/user.py#L175-L176 which defines `users`.